test/boost/aws_error_injection_test.cc

```diff
@@ -0,0 +1,240 @@
+            sstring response;
+            if (req._method == "POST") {
+                if (req.query_parameters.contains("uploads")) {
+                    response = "<InitiateMultipartUploadResult><Bucket>bucket</Bucket><Key>key</Key><UploadId>UploadId</UploadId></"
+                               "InitiateMultipartUploadResult>";
+                } else if (req.query_parameters.contains("uploadId")) {
+                    switch (_test_server.test_failure_policy) {
+                    case failure_policy::SUCCESS:
+                        return R"(<CompleteMultipartUploadResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/">
+                                     <Location>http://Example-Bucket.s3.Region.amazonaws.com/Example-Object</Location>
+                                     <Bucket>Example-Bucket</Bucket>
+                                     <Key>Example-Object</Key>
+                                     <ETag>"3858f62230ac3c915f300c664312c11f-9"</ETag>
+                                </CompleteMultipartUploadResult>)";
+                    case failure_policy::RETRYABLE_FAILURE:
+                        // should succeed on retry
+                        _test_server.test_failure_policy = failure_policy::SUCCESS;
+                        assert(aws::aws_error_map.at("InternalError").is_retryable());
+                        return R"(<?xml version="1.0" encoding="UTF-8"?>
+
+                                 <Error>
+                                  <Code>InternalError</Code>
+                                  <Message>We encountered an internal error. Please try again.</Message>
+                                  <RequestId>656c76696e6727732072657175657374</RequestId>
+                                  <HostId>Uuag1LuByRx9e6j5Onimru9pO4ZVKnJ2Qz7/C1NPcfTWAtRPfTaOFg==</HostId>
+                                 </Error>)";
+                    case failure_policy::NONRETRYABLE_FAILURE:
+                        assert(!aws::aws_error_map.at("InvalidAction").is_retryable());
+                        return R"(<?xml version="1.0" encoding="UTF-8"?>
+
+                                 <Error>
+                                  <Code>InvalidAction</Code>
+                                  <Message>Something went terribly wrong</Message>
+                                  <RequestId>656c76696e6727732072657175657374</RequestId>
+                                  <HostId>Uuag1LuByRx9e6j5Onimru9pO4ZVKnJ2Qz7/C1NPcfTWAtRPfTaOFg==</HostId>
+                                 </Error>)";
+                    }
+                }
+            }
+            return response;
+        }
+        server& _test_server;
+    };
+    explicit server(failure_policy _test_failure_policy) : test_failure_policy(_test_failure_policy) {
+        handler = std::make_unique<dummy_aws_request_handler>(*this);
+    }
+    future<> start() {
+        co_await seastar::async([this] {
+            uint16_t port = 2012;
```

Comment:
magic number. and even worse: this port cannot be shared by multiple servers. if we run multiple tests in `aws_error_injection_test` in parallel.

Reply:
probably could launch each test in a different namespace using `unshare`. see also `test/nodetool/suite.yaml` .